internal/providers/pke/pkeworkflow/update_master_node_group_activity.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/banzaicloud/pipeline/pkg/brn"
+	pkgCloudFormation "github.com/banzaicloud/pipeline/pkg/providers/amazon/cloudformation"
+	sdkAmazon "github.com/banzaicloud/pipeline/pkg/sdk/providers/amazon"
+	sdkCloudFormation "github.com/banzaicloud/pipeline/pkg/sdk/providers/amazon/cloudformation"
+)
+
+const UpdateMasterNodeGroupActivityName = "pke-aws-update-master-node-group"
+
+// UpdateMasterNodeGroupActivity updates a master node group.
+type UpdateMasterNodeGroupActivity struct {
+	sessionFactory AWSFactory
+	clusters       Clusters
+	tokenGenerator TokenGenerator
+
+	externalBaseURL         string
+	externalBaseURLInsecure bool
+}
+
+// UpdateMasterNodeGroupActivityInput holds the parameters for the node group update.
+type UpdateMasterNodeGroupActivityInput struct {
+	SecretID string
+	Region   string
+
+	OrganizationID uint
+
+	ClusterName string
+	ClusterID   uint
+
+	StackName string
+
+	NodePoolName    string
+	NodePoolVersion string
+
+	NodeVolumeSize  int
+	NodeImage       string
+	Version         string
+	DesiredCapacity int64
+
+	ClusterTags map[string]string
+}
+
+type UpdateMasterNodeGroupActivityOutput struct {
+	NodePoolChanged bool
+}
+
+// NewUpdateMasterNodeGroupActivity creates a new UpdateMasterNodeGroupActivity instance.
+func NewUpdateMasterNodeGroupActivity(
+	sessionFactory AWSFactory,
+	clusters Clusters,
+	tokenGenerator TokenGenerator,
+	externalBaseURL string,
+	externalBaseURLInsecure bool,
+) UpdateMasterNodeGroupActivity {
+	return UpdateMasterNodeGroupActivity{
+		sessionFactory:          sessionFactory,
+		clusters:                clusters,
+		tokenGenerator:          tokenGenerator,
+		externalBaseURL:         externalBaseURL,
+		externalBaseURLInsecure: externalBaseURLInsecure,
+	}
+}
+
+// Register registers the activity in the worker.
+func (a UpdateMasterNodeGroupActivity) Register() {
+	activity.RegisterWithOptions(a.Execute, activity.RegisterOptions{Name: UpdateMasterNodeGroupActivityName})
+}
+
+// Execute is the main body of the activity, returns true if there was any update and that was successful.
+func (a UpdateMasterNodeGroupActivity) Execute(ctx context.Context, input UpdateMasterNodeGroupActivityInput) (UpdateMasterNodeGroupActivityOutput, error) {
+	providerSecret, err := brn.Parse(input.SecretID)
+	if err != nil {
+		return UpdateMasterNodeGroupActivityOutput{}, errors.WrapIf(err, "failed to parse secret BRN")
+	}
+	sess, err := a.sessionFactory.New(providerSecret.OrganizationID, providerSecret.ResourceID, input.Region)
+	if err = errors.WrapIf(err, "failed to create AWS session"); err != nil { // internal error?
+		return UpdateMasterNodeGroupActivityOutput{}, err
+	}
+
+	cluster, err := a.clusters.GetCluster(ctx, input.ClusterID)
+	if err != nil {
+		return UpdateMasterNodeGroupActivityOutput{}, err
+	}
+
+	var isMultiMaster bool
+	for _, np := range cluster.GetNodePools() {
+		if input.NodePoolName == np.Name && np.Master {
+			isMultiMaster = np.Count > 1
```

Comment:
do we want to continue with single-master upgrades?

Reply:
Good question, I haven't tried multi-master TBH, and the current enterprise counterpart of this won't work for sure.

Reply:
So we probably should prohibit it.